File: frontend/src/types/workout.ts
export type TrainingGoal =
  | "fat_loss"
  | "muscle_gain"
  | "strength"
  | "general_fitness"
  | "endurance"
  | "football_performance"
  | "goalkeeper_performance";

export interface WorkoutExercise {
  exerciseId: string;
  name: string;
  description: string;
  muscleGroups: string[];
  equipment: string[];
  sets: number;
  reps: string;
  restSeconds: number;
  tempo: string;
  notes: string;
}

export interface WorkoutDay {
  id: string;
  dayNumber: number;
  title: string;
  focus: string;
  estimatedDurationMinutes: number;
  exercises: WorkoutExercise[];
}

export interface WorkoutPlan {
  id: string;
  goal: TrainingGoal;
  experience: "beginner" | "intermediate" | "advanced";
  location: "commercial_gym" | "home_gym" | "bodyweight_only";
  equipment: string[];
  injuries: string[];
  availableDays: number;
  sessionDurationMinutes: number;
  days: WorkoutDay[];
  status: "active" | "archived";
  version: number;
  generatedAt: string;
}

export interface ExerciseProgress {
  exerciseId: string;
  completedSets: number;
  skipped: boolean;
}

export interface WorkoutSession {
  id: string;
  planId: string;
  workoutDayId: string;
  status: "in_progress" | "completed";
  exerciseProgress: ExerciseProgress[];
  progress: {
    completedSets: number;
    totalSets: number;
    completedExercises: number;
    totalExercises: number;
    completionPercentage: number;
  };
  startedAt: string;
  completedAt: string | null;
}

export interface CurrentWorkout {
  plan: WorkoutPlan;
  today: WorkoutDay;
  session: WorkoutSession | null;
}

export interface WorkoutHistory {
  plans: WorkoutPlan[];
  completedSessions: number;
  weeklyAdherencePercentage: number;
}

File: backend/app/schemas/workout.py
from pydantic import BaseModel, Field

from app.models.workout import WorkoutDay, WorkoutPlan, WorkoutSession


class GenerateWorkoutRequest(BaseModel):
    available_days: int | None = Field(default=None, ge=2, le=6)
    session_duration_minutes: int | None = Field(default=None, ge=30, le=90)


class UpdateExerciseRequest(BaseModel):
    completed_sets: int = Field(ge=0, le=10)
    skipped: bool = False


class CurrentWorkoutResponse(BaseModel):
    plan: WorkoutPlan
    today: WorkoutDay
    session: WorkoutSession | None = None


class WorkoutHistoryResponse(BaseModel):
    plans: tuple[WorkoutPlan, ...]
    completed_sessions: int = Field(ge=0)
    weekly_adherence_percentage: int = Field(ge=0, le=100)
